galerautils/src/gu_enc_mmap.cpp

```diff
@@ -0,0 +1,628 @@
+#include "gu_enc_mmap.hpp"
+#include "gu_throw.hpp"
+#include "gu_logger.hpp"
+#include "gu_enc_debug.hpp"
+#include "gu_enc_utils.hpp"
+#include "gu_enc_pmemory_manager.hpp"
+#include "gu_enc_pmemory_manager_pool.hpp"
+
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <signal.h>
+#include <sys/mman.h>
+
+#include <cassert>
+#include <thread>
+#include <set>
+
+
+namespace gu {
+
+#define IS_LAST_PAGE(_page) (_page == pagesCnt_-1)
+
+#define REAL_ENCRYPTION 1
+
+// maximum number of PMemoryManagers waiting in the pool
+static const size_t MANAGERS_POOL_SIZE = 10;
+PMemoryManagerPool memoryManagerPool(MANAGERS_POOL_SIZE);
+
+// EncMMap objects repository
+static std::atomic_flag encMmapsLock = ATOMIC_FLAG_INIT;
+
+struct EncMMapDescriptor {
+    unsigned char*   start_;
+    unsigned char*   end_;
+};
+std::map<EncMMap*, EncMMapDescriptor> encMMaps;
+
+static void addEncMMap(EncMMap *mmap, unsigned char* ptr, size_t size) {
+    while (encMmapsLock.test_and_set(std::memory_order_acquire)) {
+        std::this_thread::yield();
+    }
+    encMMaps[mmap] = {ptr, ptr+size};
+    encMmapsLock.clear(std::memory_order_release);
+}
+
+static void delEncMMap(EncMMap *mmap) {
+    while (encMmapsLock.test_and_set(std::memory_order_acquire)) {
+        std::this_thread::yield();
+    }
+    encMMaps.erase(mmap);
+    encMmapsLock.clear(std::memory_order_release);
+}
+
+static EncMMap* getEncMMap(unsigned char* ptr) {
+    for (auto m : encMMaps) {
```

Comment:
It is not missing. It is done this way to not lock conflicting signals. Please check the logic and the detailed explanation in `signal_handler` function.

Reply:
Yes, I see now. Then I would do something here to make the API cleaner - because right now, other functions do locking, so this is different.

Maybe this function could be tryGetEncMMap instead, and return null if it can't get the lock?

Reply:
done